### PR TITLE
drivers/dose: make use of start condition received interrupt 

### PR DESCRIPTION
**Pull request by benpicco**
### Contribution description

DOSE can make use of a GPIO to detect the start condition, however some UARTs already have this capability build in and can generate an interrupt when a start condition is detected.

Implement this feature for the sam0 family and make use of it in the DOSE driver.

### Testing procedure

DOSE should perform just as well without the extra wire / GPIO with this feature enabled. (not sure how to verify this)

### CODE REVIEW

**Comment by MrKevinWeiss** [comment]
Should this be in this release? If so maybe you can unicast some people for a review and add some more info on how to test.

**Comment by jue89** [comment]
I think, I have the required hardware at hand for testing. But I'm not able to test before 6th of July.

**Comment by MrKevinWeiss** [comment]
Yikes, 1 day before hard feature freeze. @benpicco can this wait until a later release?

**Comment by fjmolinas** [comment]
> I think, I have the required hardware at hand for testing. But I'm not able to test before 6th of July.

Maybe @jue89 might test soon now :)

**Comment by jue89** [comment]
Thank you very much for improving DOSE and for your patience! I finally found some time to have look into your patch.

First of all: I've integrated the `samr30-xpro` into the DOSE net that already drives my home automation stuff.

I've modified the board configuration to switch from the internal RF interface to DOSE:

<details>

```diff
diff --git a/boards/samr30-xpro/Makefile.dep b/boards/samr30-xpro/Makefile.dep
index 4224c14c79..daafd22273 100644
--- a/boards/samr30-xpro/Makefile.dep
+++ b/boards/samr30-xpro/Makefile.dep
@@ -1,5 +1,5 @@
 ifneq (,$(filter netdev_default,$(USEMODULE)))
-  USEMODULE += at86rf212b
+  USEMODULE += dose
 endif
 
 ifneq (,$(filter saul_default,$(USEMODULE)))
diff --git a/boards/samr30-xpro/include/periph_conf.h b/boards/samr30-xpro/include/periph_conf.h
index b5b2add878..74a0cdf1fb 100644
--- a/boards/samr30-xpro/include/periph_conf.h
+++ b/boards/samr30-xpro/include/periph_conf.h
@@ -70,11 +70,22 @@ static const uart_conf_t uart_config[] = {
         .tx_pad   = UART_PAD_TX_0,
         .flags    = UART_FLAG_NONE,
         .gclk_src = SAM0_GCLK_MAIN,
+    },
+        {    /* EXT1 & EXT3 Pin Header */
+        .dev      = &(SERCOM1->USART),
+        .rx_pin   = GPIO_PIN(PA, 17),
+        .tx_pin   = GPIO_PIN(PA, 16),
+        .mux      = GPIO_MUX_C,
+        .rx_pad   = UART_PAD_RX_1,
+        .tx_pad   = UART_PAD_TX_0,
+        .flags    = UART_FLAG_NONE,
+        .gclk_src = SAM0_GCLK_MAIN,
     }
 };
 
 /* interrupt function name mapping */
 #define UART_0_ISR          isr_sercom0
+#define UART_1_ISR          isr_sercom1
 
 #define UART_NUMOF          ARRAY_SIZE(uart_config)
 /** @} */
@@ -127,15 +138,6 @@ static const spi_conf_t spi_config[] = {
  * @{
  */
 static const i2c_conf_t i2c_config[] = {
-    {    /* EXT1 & EXT3 Pin Header */
-        .dev      = &(SERCOM1->I2CM),
-        .speed    = I2C_SPEED_NORMAL,
-        .scl_pin  = GPIO_PIN(PA, 17),
-        .sda_pin  = GPIO_PIN(PA, 16),
-        .mux      = GPIO_MUX_C,
-        .gclk_src = SAM0_GCLK_MAIN,
-        .flags    = I2C_FLAG_NONE
-    }
 };
 #define I2C_NUMOF          ARRAY_SIZE(i2c_config)
 /** @} */
```

</details>

Some ping testing:

```
> ping ff02::1
2021-07-16 13:25:18,904 # ping ff02::1
2021-07-16 13:25:18,935 # 12 bytes from fe80::e449:92ff:fe1a:65f9%5: icmp_seq=0 ttl=64 time=23.336 ms
2021-07-16 13:25:18,950 # 12 bytes from fe80::5ca1:1dff:fe36:b359%5: icmp_seq=0 ttl=64 time=37.887 ms (DUP!)
2021-07-16 13:25:18,967 # 12 bytes from fe80::bcf3:1dff:fe6e:d30f%5: icmp_seq=0 ttl=64 time=55.049 ms (DUP!)
2021-07-16 13:25:18,981 # 12 bytes from fe80::d840:cbff:fee8:83c1%5: icmp_seq=0 ttl=64 time=68.777 ms (DUP!)
2021-07-16 13:25:19,037 # 12 bytes from fe80::9493:c4ff:fe86:83b1%5: icmp_seq=0 ttl=64 time=125.248 ms (DUP!)
2021-07-16 13:25:19,936 # 12 bytes from fe80::d840:cbff:fee8:83c1%5: icmp_seq=1 ttl=64 time=24.104 ms
2021-07-16 13:25:19,954 # 12 bytes from fe80::bcf3:1dff:fe6e:d30f%5: icmp_seq=1 ttl=64 time=41.261 ms (DUP!)
2021-07-16 13:25:19,968 # 12 bytes from fe80::5ca1:1dff:fe36:b359%5: icmp_seq=1 ttl=64 time=55.440 ms (DUP!)
2021-07-16 13:25:19,983 # 12 bytes from fe80::e449:92ff:fe1a:65f9%5: icmp_seq=1 ttl=64 time=70.096 ms (DUP!)
2021-07-16 13:25:20,037 # 12 bytes from fe80::9493:c4ff:fe86:83b1%5: icmp_seq=1 ttl=64 time=124.619 ms (DUP!)
2021-07-16 13:25:20,937 # 12 bytes from fe80::bcf3:1dff:fe6e:d30f%5: icmp_seq=2 ttl=64 time=24.374 ms
2021-07-16 13:25:20,937 # 
2021-07-16 13:25:20,940 # --- ff02::1 PING statistics ---
2021-07-16 13:25:20,947 # 3 packets transmitted, 3 packets received, 8 duplicates, 0% packet loss
2021-07-16 13:25:20,951 # round-trip min/avg/max = 23.336/59.108/125.248 ms
```

I've used printf debugging to verify that the start of a frame is actually sensed.

Your patch works as advertised! :)

File: drivers/include/periph/uart.h
Comment (jue89) [inline drivers/include/periph/uart.h:270]: I think we should use a more verbose name here. Users stumbling over this function won't be able to tell what it does by looking at its name. Maybe something like `uart_rxstart_irq_configure()`? Maybe someone else has a better idea.

File: drivers/include/periph/uart.h
Comment (jue89) [inline drivers/include/periph/uart.h:277]: Maybe `uart_rxstart_irq_enable()`?

File: drivers/include/periph/uart.h
Comment (jue89) [inline drivers/include/periph/uart.h:284]: Maybe `uart_rxstart_irq_disable()`?

File: cpu/sam0_common/periph/uart.c
Comment (jue89) [inline cpu/sam0_common/periph/uart.c:429]: Can this become a race like in #16478 if `dev(uartnum)->INTENSET.bit.RXS` has been cleared between raising the IRQ and handling the IRQ?

File: tests/periph_uart/main.c
Comment (jue89) [inline tests/periph_uart/main.c:357]: I really like this `test` command! It really help during setting up the board config for a second UART.

**Review by jue89 — CHANGES_REQUESTED** [review]
First round of review comments:

File: cpu/sam0_common/periph/uart.c
Comment (benpicco) [reply cpu/sam0_common/periph/uart.c:429]: I don't see how this could cause an issue.

We could of course always clear all flags, but for the collision detection feature it would be handy if the collision bit remains untouched by the interrupt handler and we can simply read it out after `uart_write()`

**Comment by jue89** [comment]
I've tried the ping experiment for 20000 ping requests:

```
(snip)
2021-07-16 17:03:01,388 # --- ff02::1 PING statistics ---
2021-07-16 17:03:01,396 # 20000 packets transmitted, 19999 packets received, 79702 duplicates, 0% packet loss
2021-07-16 17:03:01,401 # round-trip min/avg/max = 21.872/60.553/250.598 ms
```

File: drivers/include/periph/uart.h
Comment (benpicco) [reply drivers/include/periph/uart.h]: Ok, changed the name.

File: cpu/sam0_common/periph/uart.c
Comment (jue89) [reply cpu/sam0_common/periph/uart.c:430]: Okay, I just found the fallowing paragraph in the SAML21 datasheet:

> An interrupt request is generated when the interrupt flag is set and if the corresponding interrupt is enabled. The
> interrupt request remains active until either the interrupt flag is cleared, the interrupt is disabled, or the USART is
> reset.

Disabling the IRQ will stop triggering the ISR over and over again (which wasn't the case for the STM32).

> We could of course always clear all flags, but for the collision detection feature it would be handy if the collision bit remains untouched by the interrupt handler and we can simply read it out after uart_write()

How will this improve the collision detection? I think we can't distinguish between an foreign or our start bit. The TXSTART IRQ would be triggered either way.

File: drivers/include/periph/uart.h
Comment (jue89) [reply drivers/include/periph/uart.h]: Have you omitted the `_irq` intentionally? (`uart_rxstart_irq_configure()` vs. `uart_rxstart_configure()`)

File: cpu/sam0_common/Kconfig
Comment (jue89) [inline cpu/sam0_common/Kconfig:24]: I think this is more descriptive:

```suggestion
    select HAS_PERIPH_UART_RXSTART_IRQ
```

File: cpu/sam0_common/Makefile.features
Comment (jue89) [inline cpu/sam0_common/Makefile.features:22]: I think this is more descriptive.

```suggestion
FEATURES_PROVIDED += periph_uart_rxstart_irq
```

File: drivers/include/periph/uart.h
Comment (jue89) [inline drivers/include/periph/uart.h:265]:
```suggestion
 *
 * @note    You have to add the module `periph_uart_rxstart_irq` to your project
 *          to enable this function
 *
```

File: drivers/include/periph/uart.h
Comment (jue89) [inline drivers/include/periph/uart.h:274]:
```suggestion
 *
 * @note    You have to add the module `periph_uart_rxstart_irq` to your project
 *          to enable this function
 *
```

File: drivers/include/periph/uart.h
Comment (jue89) [inline drivers/include/periph/uart.h:281]:
```suggestion
 *
 * @note    You have to add the module `periph_uart_rxstart_irq` to your project
 *          to enable this function
 *
```

File: drivers/include/periph/uart.h
Comment (benpicco) [reply drivers/include/periph/uart.h]: I just wanted to keep it short, but can add the `_irq`

**Review by jue89 — CHANGES_REQUESTED** [review]
I think this will improve intelligibility for first-time readers stumbling over the this UART feature.

File: drivers/include/periph/uart.h
Comment (jue89) [reply drivers/include/periph/uart.h]: Even though I never programmed Java and never experienced the common spirit of using sentences for function names, I started to prefer the descriptive names over short ones :D

But I'm okay with the short version, as well. Maybe someone else has another opinion on that?!

**Comment by jue89** [comment]
If you think, everything has been addressed: please squash!

File: cpu/sam0_common/periph/uart.c
Comment (benpicco) [reply cpu/sam0_common/periph/uart.c:430]: Collision detection is a separate feature of the UART unrelated to this PR (`COLDEN` / `STATUS.COLL`)

Comment (benpicco) [reply cpu/sam0_common/periph/uart.c:430]: I'm silly, this already gets cleared on line 419

File: drivers/include/periph/uart.h
Comment (jue89) [inline drivers/include/periph/uart.h:293]: I would align this function name to `uart_rxstart_irq_configure()`:

```suggestion
void uart_rxstart_irq_disable(uart_t uart);
```

File: cpu/sam0_common/periph/uart.c
Comment (jue89) [inline cpu/sam0_common/periph/uart.c:391]: I think, we should clear pending IRQs. If I understand the datasheet correctly, the ISR would fire immediately after enabling the IRQ.

File: cpu/sam0_common/periph/uart.c
Comment (jue89) [inline cpu/sam0_common/periph/uart.c:409]: Unrelated change?

**Review by jue89 — CHANGES_REQUESTED** [review]
After going through your code a second time, I spot a few minor things:

File: cpu/sam0_common/periph/uart.c
Comment (jue89) [reply cpu/sam0_common/periph/uart.c:409]: Ah I see. You are clearing pending IRQs in the first lines of the ISR ... sorry I haven't seen this.

File: cpu/sam0_common/periph/uart.c
Comment (jue89) [reply cpu/sam0_common/periph/uart.c:391]: Since you are clearing pending IRQs in the first lines of the ISR, this shouldn't be necessary, right?

**Comment by jue89** [comment]
I just started another ping test and it got stuck ... I guess it was tripped by xtimer. Maybe the timer back-off is too short for the high IRQ rate? I try to reproduce this issue and adjust the back-off.

```
2021-07-27 20:23:18,190 # 12 bytes from fe80::e449:92ff:fe1a:65f9%5: icmp_seq=846 ttl=64 time=22.442 ms
2021-07-27 20:23:18,205 # 12 bytes from fe80::bcf3:1dff:fe6e:d30f%5: icmp_seq=846 ttl=64 time=36.646 ms (DUP!)
2021-07-27 20:23:18,220 # 12 bytes from fe80::d840:cbff:fee8:83c1%5: icmp_seq=846 ttl=64 time=51.915 ms (DUP!)
2021-07-27 20:23:18,238 # 12 bytes from fe80::5ca1:1dff:fe36:b359%5: icmp_seq=846 ttl=64 time=69.744 ms (DUP!)
2021-07-27 20:23:18,263 # 12 bytes from fe80::9493:c4ff:fe86:83b1%5: icmp_seq=846 ttl=64 time=95.154 ms (DUP!)
2021-07-27 20:23:18,391 # 12 bytes from fe80::5ca1:1dff:fe36:b359%5: icmp_seq=847 ttl=64 time=23.522 ms
2021-07-27 20:23:18,407 # 12 bytes from fe80::d840:cbff:fee8:83c1%5: icmp_seq=847 ttl=64 time=38.974 ms (DUP!)
2021-07-27 20:23:18,422 # 12 bytes from fe80::bcf3:1dff:fe6e:d30f%5: icmp_seq=847 ttl=64 time=53.607 ms (DUP!)
2021-07-27 20:23:18,438 # 12 bytes from fe80::e449:92ff:fe1a:65f9%5: icmp_seq=847 ttl=64 time=70.046 ms (DUP!)
2021-07-27 20:23:18,493 # 12 bytes from fe80::9493:c4ff:fe86:83b1%5: icmp_seq=847 ttl=64 time=124.669 ms (DUP!)
2021-07-27 20:23:32,907 # gnrc_netif: possibly lost interrupt.
2021-07-27 20:23:33,321 # gnrc_netif: possibly lost interrupt.
2021-07-27 20:23:37,642 # gnrc_netif: possibly lost interrupt.
```

File: cpu/sam0_common/periph/uart.c
Comment (jue89) [inline cpu/sam0_common/periph/uart.c:423]: This part clears IRQ flags that aren't handled by this ISR. We should mask-out those flags.

**Review by jue89 — CHANGES_REQUESTED** [review]
I think, I found the problem.

GDB to the rescue:
```
(gdb) info threads
  Id   Target Id                                 Frame 
  1    Thread 1 (Name: main, Pending)            __ISB () at /home/jue/Projects/RIOT/cpu/cortexm_common/include/vendor/cmsis_gcc.h:868
  2    Thread 2 (Name: pktdump, Blocked receive) __ISB () at /home/jue/Projects/RIOT/cpu/cortexm_common/include/vendor/cmsis_gcc.h:868
  3    Thread 3 (Name: ipv6, Pending)            __ISB () at /home/jue/Projects/RIOT/cpu/cortexm_common/include/vendor/cmsis_gcc.h:868
  4    Thread 4 (Name: udp, Blocked receive)     __ISB () at /home/jue/Projects/RIOT/cpu/cortexm_common/include/vendor/cmsis_gcc.h:868
* 5    Thread 5 (Name: dose, Running)            uart_write (uart=<optimized out>, data=0x20001438 <_netdev_eth_stack+663> "\340\026", data@entry=0x20001437 <_netdev_eth_stack+662> "", len=len@entry=1) at /home/jue/Projects/RIOT/cpu/sam0_common/periph/uart.c:69
(gdb) bt
#0  uart_write (uart=<optimized out>, data=0x20001438 <_netdev_eth_stack+663> "\340\026", data@entry=0x20001437 <_netdev_eth_stack+662> "", len=len@entry=1) at /home/jue/Projects/RIOT/cpu/sam0_common/periph/uart.c:69
#1  0x0000137e in send_octet (ctx=ctx@entry=0x200016e0 <dose>, c=<optimized out>, c@entry=0 '\000') at /home/jue/Projects/RIOT/drivers/dose/dose.c:389
#2  0x000013b6 in send_data_octet (ctx=ctx@entry=0x200016e0 <dose>, c=<optimized out>) at /home/jue/Projects/RIOT/drivers/dose/dose.c:420
#3  0x00001700 in _send (dev=0x200016e0 <dose>, iolist=0x2000148c <_netdev_eth_stack+747>) at /home/jue/Projects/RIOT/drivers/dose/dose.c:449
#4  0x00005ca2 in _send (netif=0x200015a4 <_netif>, pkt=0x20001d30 <_pktbuf_buf>) at /home/jue/Projects/RIOT/sys/net/gnrc/netif/ethernet/gnrc_netif_ethernet.c:163
#5  0x000061b0 in _send (push_back=false, pkt=<optimized out>, netif=0x200015a4 <_netif>) at /home/jue/Projects/RIOT/sys/net/gnrc/netif/gnrc_netif.c:1578
#6  _gnrc_netif_thread (args=0x200015a4 <_netif>) at /home/jue/Projects/RIOT/sys/net/gnrc/netif/gnrc_netif.c:1702
#7  0x00000af8 in sched_switch (other_prio=5856) at /home/jue/Projects/RIOT/core/sched.c:263
```

The `dose` thread is hanging during `uart_write()` and don't move forward. Looking in the code reveals the only end-less loop that might trigger such a behaviour:

```c
    for (const void* end = data + len; data != end; ++data) {
        while (!dev(uart)->INTFLAG.bit.DRE) {}
        dev(uart)->DATA.reg = *data;
    }
    while (!dev(uart)->INTFLAG.bit.TXC) {}
```

I guess that clearing all IRQ flags during ISR might steal the DRE or the TCX IRQ!

**Comment by jue89** [comment]
Success! :)

```
2021-07-28 14:45:49,052 # --- ff02::1 PING statistics ---
2021-07-28 14:45:49,060 # 30000 packets transmitted, 30000 packets received, 119234 duplicates, 0% packet loss
2021-07-28 14:45:49,065 # round-trip min/avg/max = 21.702/61.183/297.718 ms
```

Please squash!

**Review by jue89 — APPROVED** [review]
ACK! Nice addition for MCUs that support start bit IRQs.

**Comment by benpicco** [comment]
Thank you for the review!

**Pull request merged**
